File: python/source/main_base.py
"""
Python 基础语法学习示例
本文件展示了 Python 的基础语法、数据结构、类和对象等核心概念
"""

# =============================================================================
# 1. 基础语法示例
# =============================================================================


def basic_syntax_examples():
    """基础语法示例函数"""

    # 变量和数据类型
    name: str = "张三"  # 字符串类型
    age: int = 25  # 整数类型
    height: float = 1.75  # 浮点数类型
    is_student: bool = True  # 布尔类型

    print("=== 基础数据类型 ===")
    print(f"姓名: {name}")
    print(f"年龄: {age}")
    print(f"身高: {height}米", f"是否学生: {is_student}")

    # 算术运算
    a: int = 10
    b: int = 3

    print("\n=== 算术运算 ===")
    print(f"{a} + {b} = {a + b}")  # 加法
    print(f"{a} - {b} = {a - b}")  # 减法
    print(f"{a} * {b} = {a * b}")  # 乘法
    print(f"{a} / {b} = {a / b:.2f}")  # 除法，保留两位小数
    print(f"{a} // {b} = {a // b}")  # 整除
    print(f"{a} % {b} = {a % b}")  # 取余
    print(f"{a} ** {b} = {a ** b}")  # 幂运算

    # 比较运算
    print("\n=== 比较运算 ===")
    print(f"{a} > {b}: {a > b}")
    print(f"{a} < {b}: {a < b}")
    print(f"{a} == {b}: {a == b}")
    print(f"{a} != {b}: {a != b}")

    # 逻辑运算
    x: bool = True
    y: bool = False

    print("\n=== 逻辑运算 ===")
    print(f"x and y: {x and y}")  # 与运算
    print(f"x or y: {x or y}")  # 或运算
    print(f"not x: {not x}")  # 非运算


# =============================================================================
# 2. 控制流示例
# =============================================================================


def control_flow_examples():
    """控制流示例函数"""

    print("\n=== if/else 条件判断 ===")

    score: int = 85

    if score >= 90:
        print("成绩优秀！")
    elif score >= 80:
        print("成绩良好！")
    elif score >= 60:
        print("成绩及格！")
    else:
        print("成绩不及格！")

    print("\n=== for 循环 ===")

    # 遍历数字范围
    print("数字 1-5:")
    for i in range(1, 6):
        print(f"  {i}")

    # 遍历列表
    fruits: list[str] = ["苹果", "香蕉", "橙子"]
    print("水果列表:")
    for fruit in fruits:
        print(f"  {fruit}")

    print("\n=== while 循环 ===")

    count: int = 3
    while count > 0:
        print(f"倒计时: {count}")
        count -= 1
    print("开始！")


# =============================================================================
# 3. 数据结构示例
# =============================================================================


def data_structure_examples():
    """数据结构示例函数"""

    print("\n=== 列表 (List) ===")

    # 列表 - 有序的可变集合
    numbers: list[int] = [1, 2, 3, 4, 5]
    print(f"原始列表: {numbers}")

    # 列表操作
    numbers.append(6)  # 添加元素
    numbers.insert(0, 0)  # 在指定位置插入
    numbers.remove(3)  # 删除元素

    print(f"修改后列表: {numbers}")
    print(f"列表长度: {len(numbers)}")
    print(f"第一个元素: {numbers[0]}")
    print(f"最后一个元素: {numbers[-1]}")

    print("\n=== 元组 (Tuple) ===")

    # 元组 - 有序的不可变集合
    coordinates: tuple[float, float] = (10.5, 20.3)
    print(f"坐标: {coordinates}")
    print(f"X坐标: {coordinates[0]}")
    print(f"Y坐标: {coordinates[1]}")

    print("\n=== 字典 (Dictionary) ===")

    # 字典 - 键值对集合
    student: dict[str, any] = {
        "姓名": "李四",
        "年龄": 20,
        "专业": "计算机科学",
        "成绩": {"数学": 90, "英语": 85},
    }

    print(f"学生信息: {student}")
    print(f"姓名: {student['姓名']}")
    print(f"年龄: {student['年龄']}")

    # 字典操作
    student["班级"] = "一班"  # 添加新键值对
    print(f"添加班级后: {student}")

    print("\n=== 集合 (Set) ===")

    # 集合 - 无序的不重复元素集合
    set1: set[int] = {1, 2, 3, 4, 5}
    set2: set[int] = {4, 5, 6, 7, 8}

    print(f"集合1: {set1}")
    print(f"集合2: {set2}")
    print(f"并集: {set1 | set2}")  # 或 set1.union(set2)
    print(f"交集: {set1 & set2}")  # 或 set1.intersection(set2)
    print(f"差集: {set1 - set2}")  # 或 set1.difference(set2)


# =============================================================================
# 4. 函数示例
# =============================================================================


def function_examples():
    """函数示例函数"""

    print("\n=== 函数定义和调用 ===")
    # 作用域示例
    name: str = "张三"

    # 简单函数
    def greet(name: str) -> str:
        """向某人问好"""
        return f"你好, {name}!"

    # 带默认参数的函数
    def calculate_area(length: float, width: float = 1.0) -> float:
        """计算矩形面积，宽度默认为1.0"""
        return length * width

    # 可变参数函数
    def sum_numbers(*numbers: int) -> int:
        """计算任意数量数字的和"""
        return sum(numbers)

    # 调用函数
    print(greet("王五"))
    print(f"矩形面积(长5宽3): {calculate_area(5, 3)}")
    print(f"矩形面积(长5默认宽): {calculate_area(5)}")
    print(f"数字和: {sum_numbers(1, 2, 3, 4, 5)}")


# =============================================================================
# 5. 类和对象示例
# =============================================================================


class Person:
    """人类 - 演示类的定义和使用"""

    def __init__(self, name: str, age: int):
        """构造函数，初始化对象属性"""
        self.name = name
        self.age = age

    def introduce(self) -> str:
        """自我介绍方法"""
        return f"我叫{self.name}，今年{self.age}岁。"

    def have_birthday(self) -> None:
        """过生日，年龄加1"""
        self.age += 1
        print(f"{self.name}过生日啦！现在{self.age}岁。")


class Student(Person):
    """学生类 - 继承自Person类"""

    def __init__(self, name: str, age: int, student_id: str):
        """学生类构造函数"""
        super().__init__(name, age)  # 调用父类构造函数
        self.student_id = student_id
        self.courses: list[str] = []

    def enroll_course(self, course_name: str) -> None:
        """选课方法"""
        self.courses.append(course_name)
        print(f"{self.student_id}选了课程: {course_name}")

    def show_courses(self) -> None:
        """显示所选课程"""
        if self.courses:
            print(f"{self.student_id}的课程: {', '.join(self.courses)}")
        else:
            print(f"{self.student_id}还没有选课")


def class_object_examples():
    """类和对象示例函数"""

    print("\n=== 类和对象 ===")

    # 创建Person对象
    person1 = Person("张三", 25)
    print(person1.introduce())

    # 创建Student对象
    student1 = Student("李四", 20, "S001")
    print(student1.introduce())

    # 调用对象方法
    student1.enroll_course("Python编程")
    student1.enroll_course("数据结构")
    student1.show_courses()

    # 过生日
    student1.have_birthday()


# =============================================================================
# 6. 异常处理示例
# =============================================================================


def exception_handling_examples():
    """异常处理示例函数"""

    print("\n=== 异常处理 ===")

    # 尝试执行可能出错的代码
    try:
        number_str = input("请输入一个数字: ")
        number = int(number_str)
        result = 100 / number
        print(f"100除以{number}的结果是: {result}")

    except ValueError:
        print("错误：请输入有效的数字！")

    except ZeroDivisionError:
        print("错误：不能除以零！")

    except Exception as e:
        print(f"发生未知错误: {e}")

    else:
        print("计算成功完成！")

    finally:
        print("异常处理示例结束。")


# =============================================================================
# 7. 异步编程示例
# =============================================================================
import asyncio, time


async def async_example():
    """简单异步示例"""

    async def order_and_serve(dish_name: str, prepare_time: int):
        await asyncio.sleep(prepare_time)
        print(f"{dish_name} 已准备完成！")
        print(f"{dish_name} 已服务完成！")

    start_time = time.time()
    print("\n=== 异步编程 ===")
    await asyncio.gather(
        order_and_serve("鱼香肉丝", 1),
        order_and_serve("红烧肉", 1),
        order_and_serve("青椒肉丝", 1),
    )
    print("异步任务完成,执行时时间: ", time.time() - start_time)


# =============================================================================
# 主函数
# =============================================================================


def main():
    """主函数 - 组织所有示例的执行"""

    print("🎯 Python 基础语法学习示例")
    print("=" * 50)

    # 按顺序执行以上7个示例
    basic_syntax_examples()
    control_flow_examples()
    data_structure_examples()
    asyncio.run(async_example())
    function_examples()
    class_object_examples()
    exception_handling_examples()

    print("\n" + "=" * 50)
    print("✅ 所有示例执行完成！")
    print("💡 建议：修改代码并运行，观察不同的输出结果")


if __name__ == "__main__":
    main()

File: c++/source/main_base.cpp
/*
 * C++ 基础语法学习示例
 * 本文件展示了 C++ 的基础语法、STL 容器、类和对象等核心概念
 * 编译: g++ -std=c++20 -o main_base main_base.cpp
 */

#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <memory>
#include <algorithm>

// =============================================================================
// 1. 基础语法示例
// =============================================================================

void basic_syntax_examples()
{
    // 变量和数据类型
    std::string name{"张三"};   // 字符串类型
    int age{25};                 // 整数类型
    double height{1.75};         // 双精度浮点
    bool is_student{true};       // 布尔类型

    std::cout << "=== 基础数据类型 ===" << std::endl;
    std::cout << "姓名: " << name << std::endl;
    std::cout << "年龄: " << age << std::endl;
    std::cout << "身高: " << height << "米" << std::endl;
    std::cout << "是否学生: " << (is_student ? "是" : "否") << std::endl;

    // 算术运算
    int a{10};
    int b{3};

    std::cout << "\n=== 算术运算 ===" << std::endl;
    std::cout << a << " + " << b << " = " << a + b << std::endl;
    std::cout << a << " - " << b << " = " << a - b << std::endl;
    std::cout << a << " * " << b << " = " << a * b << std::endl;
    std::cout << a << " / " << b << " = " << a / b << "（整数除）" << std::endl;
    std::cout << a << " / " << b << ".0 = " << static_cast<double>(a) / b << "（浮点除）" << std::endl;
    std::cout << a << " % " << b << " = " << a % b << std::endl;

    // 比较运算
    std::cout << "\n=== 比较运算 ===" << std::endl;
    std::cout << a << " > " << b << ": " << (a > b) << std::endl;
    std::cout << a << " < " << b << ": " << (a < b) << std::endl;
    std::cout << a << " == " << b << ": " << (a == b) << std::endl;
    std::cout << a << " != " << b << ": " << (a != b) << std::endl;

    // 逻辑运算
    bool x{true};
    bool y{false};

    std::cout << "\n=== 逻辑运算 ===" << std::endl;
    std::cout << "x && y: " << (x && y) << std::endl;
    std::cout << "x || y: " << (x || y) << std::endl;
    std::cout << "!x: " << (!x) << std::endl;
}

// =============================================================================
// 2. 控制流示例
// =============================================================================

void control_flow_examples()
{
    std::cout << "\n=== if/else 条件判断 ===" << std::endl;

    int score{85};

    if (score >= 90)
    {
        std::cout << "成绩优秀！" << std::endl;
    }
    else if (score >= 80)
    {
        std::cout << "成绩良好！" << std::endl;
    }
    else if (score >= 60)
    {
        std::cout << "成绩及格！" << std::endl;
    }
    else
    {
        std::cout << "成绩不及格！" << std::endl;
    }

    std::cout << "\n=== switch 语句 ===" << std::endl;

    char grade{'B'};
    switch (grade)
    {
    case 'A':
        std::cout << "优秀" << std::endl;
        break;
    case 'B':
        std::cout << "良好" << std::endl;
        break;
    case 'C':
        std::cout << "及格" << std::endl;
        break;
    default:
        std::cout << "不及格" << std::endl;
        break;
    }

    std::cout << "\n=== for 循环 ===" << std::endl;

    std::cout << "数字 1-5: ";
    for (int i{1}; i <= 5; i++)
    {
        std::cout << i << " ";
    }
    std::cout << std::endl;

    std::vector<std::string> fruits{"苹果", "香蕉", "橙子"};
    std::cout << "水果列表: ";
    for (const auto& fruit : fruits)
    {
        std::cout << fruit << " ";
    }
    std::cout << std::endl;

    std::cout << "\n=== while 循环 ===" << std::endl;

    int count{3};
    while (count > 0)
    {
        std::cout << "倒计时: " << count << std::endl;
        count--;
    }
    std::cout << "开始！" << std::endl;
}

// =============================================================================
// 3. 数据结构示例（STL 容器）
// =============================================================================

void data_structure_examples()
{
    std::cout << "\n=== std::vector（动态数组）===" << std::endl;

    std::vector<int> numbers{1, 2, 3, 4, 5};
    std::cout << "原始: ";
    for (int n : numbers)
        std::cout << n << " ";
    std::cout << std::endl;

    numbers.push_back(6);
    numbers.insert(numbers.begin(), 0);
    numbers.erase(numbers.begin() + 3);

    std::cout << "修改后: ";
    for (int n : numbers)
        std::cout << n << " ";
    std::cout << std::endl;

    std::cout << "长度: " << numbers.size() << std::endl;
    std::cout << "第一个: " << numbers.front() << std::endl;
    std::cout << "最后一个: " << numbers.back() << std::endl;

    std::cout << "\n=== std::map（有序键值对）===" << std::endl;

    std::map<std::string, int> scores;
    scores["数学"] = 90;
    scores["英语"] = 85;
    scores["编程"] = 95;

    for (const auto& [subject, score] : scores)
    {
        std::cout << subject << ": " << score << "分" << std::endl;
    }

    std::cout << "\n=== std::set（集合）===" << std::endl;

    std::set<int> set_a{1, 2, 3, 4, 5};
    std::set<int> set_b{4, 5, 6, 7, 8};

    std::cout << "集合a: ";
    for (int n : set_a)
        std::cout << n << " ";
    std::cout << std::endl;

    std::cout << "集合b: ";
    for (int n : set_b)
        std::cout << n << " ";
    std::cout << std::endl;

    // 检查元素是否存在
    std::cout << "3在a中: " << (set_a.count(3) ? "是" : "否") << std::endl;
    std::cout << "9在a中: " << (set_a.count(9) ? "是" : "否") << std::endl;

    std::cout << "\n=== std::string ===" << std::endl;

    std::string s{"Hello"};
    s += " C++";
    std::cout << "拼接: " << s << std::endl;
    std::cout << "长度: " << s.length() << std::endl;
    std::cout << "子串: " << s.substr(0, 5) << std::endl;
}

// =============================================================================
// 4. 函数示例
// =============================================================================

// 基本函数
std::string greet(const std::string& name)
{
    return "你好, " + name + "!";
}

// 带默认参数的函数
double calculate_area(double length, double width = 1.0)
{
    return length * width;
}

// 函数重载
int max_val(int a, int b) { return a > b ? a : b; }
double max_val(double a, double b) { return a > b ? a : b; }
int max_val(int a, int b, int c) { return max_val(max_val(a, b), c); }

void function_examples()
{
    std::cout << "\n=== 函数定义和调用 ===" << std::endl;

    std::cout << greet("王五") << std::endl;
    std::cout << "矩形面积(长5宽3): " << calculate_area(5, 3) << std::endl;
    std::cout << "矩形面积(长5默认宽): " << calculate_area(5) << std::endl;

    std::cout << "max(3, 5) = " << max_val(3, 5) << std::endl;
    std::cout << "max(3.1, 2.7) = " << max_val(3.1, 2.7) << std::endl;
    std::cout << "max(1, 5, 3) = " << max_val(1, 5, 3) << std::endl;

    // Lambda 表达式
    auto square = [](int x) { return x * x; };
    std::cout << "6的平方 = " << square(6) << std::endl;

    int counter{0};
    auto increment = [&counter]() { counter++; };
    increment();
    increment();
    std::cout << "counter = " << counter << std::endl;
}

// =============================================================================
// 5. 类和对象示例
// =============================================================================

class Person
{
private:
    std::string name;
    int age;

public:
    Person(std::string n, int a) : name{std::move(n)}, age{a} {}

    void introduce() const
    {
        std::cout << "我叫" << name << "，今年" << age << "岁。" << std::endl;
    }

    void have_birthday()
    {
        age++;
        std::cout << name << "过生日啦！现在" << age << "岁。" << std::endl;
    }
};

class Student : public Person
{
private:
    std::string student_id;
    std::vector<std::string> courses;

public:
    Student(std::string n, int a, std::string sid)
        : Person{std::move(n), a}, student_id{std::move(sid)} {}

    void enroll_course(const std::string& course_name)
    {
        courses.push_back(course_name);
        std::cout << student_id << "选了课程: " << course_name << std::endl;
    }

    void show_courses() const
    {
        if (courses.empty())
        {
            std::cout << student_id << "还没有选课" << std::endl;
        }
        else
        {
            std::cout << student_id << "的课程: ";
            for (size_t i{0}; i < courses.size(); i++)
            {
                std::cout << courses[i];
                if (i < courses.size() - 1)
                    std::cout << ", ";
            }
            std::cout << std::endl;
        }
    }
};

void class_object_examples()
{
    std::cout << "\n=== 类和对象 ===" << std::endl;

    Person person1("张三", 25);
    person1.introduce();
    person1.have_birthday();

    Student student1("李四", 20, "S001");
    student1.introduce();  // 继承自 Person
    student1.enroll_course("Python编程");
    student1.enroll_course("数据结构");
    student1.show_courses();
}

// =============================================================================
// 6. 异常处理示例
// =============================================================================

void exception_examples()
{
    std::cout << "\n=== 异常处理 ===" << std::endl;

    auto safe_divide = [](int a, int b) -> double
    {
        if (b == 0)
        {
            throw std::runtime_error("除数不能为零！");
        }
        return static_cast<double>(a) / b;
    };

    try
    {
        std::cout << "10 / 2 = " << safe_divide(10, 2) << std::endl;
        std::cout << "10 / 0 = " << safe_divide(10, 0) << std::endl;
    }
    catch (const std::runtime_error& e)
    {
        std::cerr << "运行时错误: " << e.what() << std::endl;
    }
    catch (const std::exception& e)
    {
        std::cerr << "标准异常: " << e.what() << std::endl;
    }

    std::cout << "异常处理完毕，程序继续运行。" << std::endl;
}

// =============================================================================
// 7. 指针示例（裸指针 - 仅供理解，实际应使用智能指针）
// =============================================================================

void pointer_examples()
{
    std::cout << "\n=== 指针基础 ===" << std::endl;

    int value{42};
    int* ptr{&value};

    std::cout << "value 的值: " << value << std::endl;
    std::cout << "value 的地址: " << &value << std::endl;
    std::cout << "ptr 存储的地址: " << ptr << std::endl;
    std::cout << "ptr 解引用的值: " << *ptr << std::endl;

    // 通过指针修改原值
    *ptr = 100;
    std::cout << "修改后 value = " << value << std::endl;

    // 指针与数组
    std::cout << "\n=== 指针与数组 ===" << std::endl;

    int arr[5]{10, 20, 30, 40, 50};
    int* arr_ptr{arr};  // 数组名即首元素地址

    std::cout << "通过指针遍历数组: ";
    for (int i{0}; i < 5; i++)
    {
        std::cout << *(arr_ptr + i) << " ";
    }
    std::cout << std::endl;

    // 动态内存分配
    std::cout << "\n=== new / delete ===" << std::endl;

    int* dyn_int{new int{999}};
    std::cout << "动态分配的整数: " << *dyn_int << std::endl;
    delete dyn_int;
    // dyn_int = nullptr;  // 防止悬垂指针

    // 空指针检查
    std::cout << "\n=== 空指针 ===" << std::endl;

    int* null_ptr{nullptr};
    if (null_ptr)
    {
        std::cout << "指针有效" << std::endl;
    }
    else
    {
        std::cout << "指针为空（nullptr）" << std::endl;
    }
}

// =============================================================================
// 主函数
// =============================================================================

int main()
{
    basic_syntax_examples();
    control_flow_examples();
    data_structure_examples();
    function_examples();
    class_object_examples();
    exception_examples();
    pointer_examples();

    std::cout << "\n✅ 所有基础示例执行完成！" << std::endl;
    std::cout << "💡 建议：修改代码并运行，观察不同的输出结果" << std::endl;

    return 0;
}